### DISCUSSION TITLE
Emitting action/filter events for api extensions

### DISCUSSION THREAD

**Discussion by br41nslug** [directus/directus#10400] [Feature Requests]
Adding the ability for extensions to emit custom events would be very useful feature for more complex extensions. For example my websocket plugin could emit an event when a client has connected or disconnected which could be consumed by other hook extensions as `action` or `filter`. This would be a nice way of keeping business logic separate from functional plugins.

On a practical note, this could be implemented as easy as exposing the existing `Emitter` instance in the `HookContext` passed to api extensions. Although for stability passing some limited version of that would be a better option as extensions definitely shouldn't be able to emit/override directus built-in events.

**Comment by rijkvanzanten** [comment]
Love it! I think exposing the emitter instance through the hook context makes the most sense to me. 

> Although for stability passing some limited version of that would be a better option as extensions definitely shouldn't be able to emit/override directus built-in events

I think this is mostly "user risk" at that point. It's similar to most other hooks honestly, if you do something silly like `process.exit(0)` in a hook, you'll cause issues in your instance. 

Another thing we'll have to document is that people don't emit the event they're currently handling, otherwise you end up in an infinite loop 🔁

**Reply by br41nslug** [reply]
Yeah preventing hooks from emitting directus events might be a bit of overkill but the infinite event loop is definitely something to watch out for!

**Comment by rijkvanzanten** [comment]
I think it might be as simple as adding `emitter` to this object:

https://github.com/directus/directus/blob/ba6f90858bce0b369aa6e6b10cfb954d9bf9c2ed/api/src/extensions.ts#L281-L288

Wanna try your hand at a PR for that @br41nslug?

Also, @nickrum are we missing anything obvious here? Wouldn't want to destroy the hooks infrastructure of course 😄

**Reply by br41nslug** [reply]
I would love to try my hand at creating a PR for this 😄

**Reply by nickrum** [reply]
Fine by me 👍

**Comment by yamiteru** [comment]
How do we use this? I tried creating an action hook and then emitting using actionEmitter but nothing happens 🤷‍♂️

**Reply by br41nslug** [reply]
For security this feature was isolated from the core event emitter so you can't use the main events to listen for these unfortunately. So using this will look something like this:
First create a hook listening for the custom event
```js
module.exports = function registerHook(_, { logger, emitter }) {
    emitter.onFilter('custom.endpoint', (data) => {
        // ... filter logic ...
        data.extra = "more info";
        data.some = "better";
        return data;
    });
};
```
Then some extension for emitting the event
```js
module.exports = function registerEndpoint(router, { emitter }) {
  router.get('/', async (req, res) => {
    const data = await emitter.emitFilter('custom.endpoint', { "some": "cool", "data": 42 });
    res.send(JSON.stringify(data));
  });
}
```

**Comment by rijkvanzanten** [comment]
Heya! Thank you for taking the time to submit this request! 

It has been over **90 days**, and this discussion has not received at least **15 votes** from the community. This means that we don't feel like there's enough community interest to warrant further R&D into this topic at this time. :ice_cube:

This request will now be closed to keep our discussions tidy. Please reach out if you have any questions!

_For more information, see our [Feature Request Process](https://docs.directus.io/contributing/feature-request-process.html)._
